gt7communication.py

```diff
@@ -167,7 +178,58 @@ def __init__(self, playstation_ip):
         self.current_lap = Lap()
         self.session = Session()
         self.laps = []
+        self.data_history = deque(maxlen = 20)
         self.last_data = GTData(None)
+        self.state = None
+
+    def log_state(self, state):
+        match state:
+            case None:
+                log.info('Stream State: NONE')
+            case StreamState.MENU:
+                log.info('Stream State: MENU')
+            case StreamState.PRERACE:
+                log.info('Stream State: PRERACE')
+            case StreamState.LIVE:
+                log.info('Stream State: LIVE')
+            case StreamState.REPLAY:
+                log.info('Stream State: REPLAY')
+
+    def check_for_replay_start(self, data_history):
+        for index in range (0, len(data_history) - (1 + 5 + 1)):
+            if data_history[index].total_laps >= 0 and data_history[index + 1].total_laps < 0:
+                # no more than 5 data packets and 5 frames, usually 3 packets
+                for index_ in range(index + 2, index + 7):
+                    if data_history[index_].total_laps >= 0 and (data_history[index].recvtime - data_history[index_].recvtime) <= 5.0 / 60:
+                        for index__ in range (index + 1, len(data_history)):
+                            data_history.pop()
+                        return True
+        return False
+
+    def check_for_replay_end(self, data_history):
```

Comment:
Could you explain what the logic behind this function and the function below is? Some magic numbers and logic that will be hard to maintain in the future.

I would prefer a manual approach of recording and not recording laps.

Reply:
Sorry, I only just saw your comment.
`data_history` is a FIFO queue of packets.
At replay start, `total_laps` will quickly flip to -1 for a fraction of a second, it usually lasts 3 packets, but just for safety, I assumed it could last a bit longer. `check_for_replay_start()` just implements this logic, it just looks for this pattern `… >=0, -1, >=0 …`. It starts by scanning the latest and its previous 5 packets, when `total_laps` is no less than 0, and its previous `total_laps` is -1, replay has probably started; then we scan further back, and look for a `total_laps` value no less than 0, within 5 packets or 5 frames worth of packets, whichever is less; if we found one, then we can determine that replay has started, then we discard all packets before the replay starts.
At replay end, we look for an abrupt change in engine RPM. In `check_for_replay_end()`, if `rpm` changed more than 10% within 1/30 of a second, we determine that the replay has ended, then discard all packets before the replay end.
There are magic numbers, but they can be easily extracted and renamed.